alf/optimizers/traj_optimizers.py

```diff
@@ -89,3 +91,115 @@ def obtain_solution(self, time_step: TimeStep, state):
+            epsilon (float): a minimum variance threshold. If the variance of
+                the population falls below it, the CEM iteration will stop.
+            tau (float): a value in (0, 1) for softly updating the population
+                mean and variance:
+                    mean = (1 - tau) * mean + tau * new_mean
+                    var = (1 - tau) * var + tau * new_var
+            min_var (float): minimum value of the variance for the Gaussian
+                distribution to sample from
+        """
+        super().__init__()
+        self._solution_dim = solution_dim
+        self._population_size = population_size
+        self._upper_bound = upper_bound
+        self._lower_bound = lower_bound
+        self._elite_size = elite_size
+        self._max_iter_num = max_iter_num
+        self._epsilon = epsilon
+        self._tau = tau
+        self._min_var = min_var
+
+    def obtain_solution(self,
+                        time_step: TimeStep,
+                        state,
+                        init_mean=None,
+                        init_var=None):
+        """Minimize the cost function provided by using the CEM method.
+
+        Args:
+            time_step (TimeStep): the initial time_step to start rollout
+            state: the initial state to start rollout
+            init_mean (None|Tensor): initial mean of the population
+            init_var (None|Tensor): initial variance of the population
+        """
+        init_obs = time_step.observation
+        batch_size = init_obs.shape[0]
+
+        if init_mean is None:
+            # [B, 1, solution_dim]
+            init_mean = torch.zeros(batch_size, 1, self._solution_dim)
+        else:
+            assert init_mean.shape == (batch_size, 1, self._solution_dim)
+
+        if init_var is None:
+            init_var = torch.ones(batch_size, 1, self._solution_dim)
+        else:
+            assert init_var.shape == (batch_size, 1, self._solution_dim)
+
+        i = 0
+        pop_mean = init_mean
+        pop_var = init_var
+
+        # [B, population, solution_dim]
+        trunc_normal_samples = torch.zeros(batch_size, self._population_size,
+                                           self._solution_dim)
+
+        while i < self._max_iter_num and pop_var.view(
+                -1).max() > self._epsilon:
+            # compute the distance to lower and upper bound
+            distance_to_lb = pop_mean - self._lower_bound
+            distance_to_ub = self._upper_bound - pop_mean
+
+            # compute the constrained var based on the computed distance
+            constrained_var = torch.min(
+                torch.min((distance_to_lb / 2.0)**2, (distance_to_ub / 2.0)
+                          **2), pop_var)
+            constrained_var = constrained_var.clamp(min=self._min_var)
+
+            trunc_normal_samples = alf.initializers.trunc_normal_(
+                trunc_normal_samples)
```

Comment:
It's better to modify `trunc_normal_()` to handle vector form of mean, std, a, b.

Reply:
Leave ```trunc_normal_``` unchanged for now as it is a pytorch function (copied from a higher version). Will think about proper ways to deal with this issue.

Reply:
Changed to using Normal as generation distribution. Performs similar to TruncNormal but computation speed is much faster.